policies/recipes/steps.go

```diff
@@ -90,8 +99,344 @@ func parsePermissions(s string) (os.FileMode, error) {
+		if err == nil {
+			_, err = io.Copy(dst, reader)
+		}
+		reader.Close()
+
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.ModTime())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func extractTar(tarName string, dst string) error {
+	file, err := os.Open(tarName)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(file)
+
+	// Check for conflicts
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag == tar.TypeDir && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	file.Seek(0, 0)
+	tr = tar.NewReader(file)
+
+	// Create dirs
+	for {
+		var err error
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag != tar.TypeDir {
+			continue
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		err = os.MkdirAll(filen, os.FileMode(header.Mode))
+		if err != nil {
+			return err
+		}
+		err = os.Chown(filen, header.Uid, header.Gid)
+		if err != nil {
+			return err
+		}
+	}
+
+	file.Seek(0, 0)
+	tr = tar.NewReader(file)
+
+	// Create files.
+	for {
+		var err error
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if err != nil {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		switch header.Typeflag {
+		case tar.TypeDir:
+			continue
+		case tar.TypeReg, tar.TypeRegA:
+			fmt.Printf("os.Create %s (owner %s/%d group %s/%d)\n", filen, header.Uname, header.Uid, header.Gname, header.Gid)
+			dst, err := os.Create(filen)
+			if err == nil {
+				_, err = io.Copy(dst, tr)
+			}
+		case tar.TypeLink:
+			err = os.Link(header.Linkname, filen)
+			continue
+		case tar.TypeSymlink:
+			err = os.Symlink(header.Linkname, filen)
+			continue
+		case tar.TypeChar:
+			err = mkCharDevice(filen, uint32(header.Devmajor), uint32(header.Devminor))
+		case tar.TypeBlock:
+			err = mkBlockDevice(filen, uint32(header.Devmajor), uint32(header.Devminor))
+		case tar.TypeFifo:
+			err = mkFifo(filen, uint32(header.Mode))
+		default:
+			fmt.Printf("unknown type for %s\n", filen)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		err = os.Chmod(filen, os.FileMode(header.Mode))
```

Comment:
Why chmod here instead of in os.OpenFile?

Reply:
I could have the openfile too but the chmod here also works on block and char devices. I'll move it so it's handled in each file type.